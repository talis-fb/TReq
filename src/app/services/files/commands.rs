use std::path::PathBuf;

use tokio::sync::oneshot;

use super::service::FileServiceInstance;
use crate::app::service_commands::Command;

pub type CommandFileService<Resp> = Command<FileServiceInstance, Resp>;

pub struct CommandsFactory;

impl CommandsFactory {
    pub fn get_or_create_data_file(path: String) -> CommandFileService<PathBuf> {
        let (tx, rx) = oneshot::channel();

        Command::from(|service: FileServiceInstance| {
            if let Ok(path_file) = service.get_or_create_data_file(path) {
                tx.send(path_file).ok();
            }
            service
        })
        .with_response(rx)
    }

    pub fn get_or_create_config_file(path: String) -> CommandFileService<PathBuf> {
        let (tx, rx) = oneshot::channel();

        Command::from(|service: FileServiceInstance| {
            if let Ok(path_file) = service.get_or_create_config_file(path) {
                tx.send(path_file).ok();
            }

            service
        })
        .with_response(rx)
    }

    pub fn get_or_create_temp_file(path: String) -> CommandFileService<PathBuf> {
        let (tx, rx) = oneshot::channel();

        Command::from(|service: FileServiceInstance| {
            if let Ok(path_file) = service.get_or_create_temp_file(path) {
                tx.send(path_file).ok();
            }
            service
        })
        .with_response(rx)
    }

    pub fn find_all_data_files() -> CommandFileService<Vec<PathBuf>> {
        let (tx, rx) = oneshot::channel();

        Command::from(|service: FileServiceInstance| {
            if let Ok(path_file) = service.find_all_data_files() {
                tx.send(path_file).ok();
            }
            service
        })
        .with_response(rx)
    }

    pub fn remove_file(path: PathBuf) -> CommandFileService<()> {
        Command::from(|service: FileServiceInstance| {
            service.remove_file(path).ok();
            service
        })
    }
}
