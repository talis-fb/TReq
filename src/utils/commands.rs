use std::ops::FnOnce;

pub type CommandClosure<Facade> =
    dyn FnOnce(Box<Facade>) -> Result<Box<Facade>, String> + Send + Sync;
    
pub type Command<Facade> = Box<CommandClosure<Facade>>;

pub fn from<Facade, Param, Return>(cl: impl FnOnce(Param) -> Return) -> Command<Facade> {
    todo!()
}
