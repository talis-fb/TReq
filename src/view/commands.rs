use std::io::{stderr, stdout};

use async_trait::async_trait;

use crate::app::provider::Provider;
use crate::app::services::request::entity::RequestData;
use crate::view::cli::writer::CrosstermCliWriter;
use crate::view::commands::submit_request::BasicRequestExecutor;

pub mod submit_request;

#[async_trait]
pub trait AppCommandExecutor {
    async fn execute(&mut self, provider: Box<dyn Provider + Send>) -> anyhow::Result<()>;
}

#[derive(Debug, PartialEq, Eq)]
pub enum AppCommand {
    /// A basic GET request
    BasicRequest { req: RequestData },
}

impl AppCommand {
    pub fn get_executor(&self) -> Box<dyn AppCommandExecutor> {
        let writer_stderr = CrosstermCliWriter {
            stdout: Box::new(stderr()),
        };
        let writer_stdout = CrosstermCliWriter {
            stdout: Box::new(stdout()),
        };

        use AppCommand as E;
        match self {
            E::BasicRequest { req } => Box::new(BasicRequestExecutor {
                req: req.clone(),
                writer_stdout,
                writer_stderr,
            }),
        }
    }
}
