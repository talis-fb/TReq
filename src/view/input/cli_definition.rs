#![allow(dead_code)]
use clap::{command, Arg, ArgAction, Command};

pub fn root_command() -> Command {
    let mut app = command!();

    for method in ["GET", "POST", "PUT", "DELETE", "HEAD", "PATCH"] {
        app = app.subcommand(
            Some(
                Command::new(method)
                    .about(format!("Does a {method} request"))
                    .override_usage(format!("treq {method} <URL> [OPTIONS + REQUEST_ITENS... ]"))
                    .arg(Arg::new("inputs").value_name("URL").required(true)),
            )
            .map(add_request_items_args)
            .map(add_raw_flag)
            .map(add_save_as_flag)
            .unwrap(),
        )
    }

    app = {
        app.subcommand(
            Some(
                Command::new("run")
                    .override_usage(format!(
                        "treq run <REQUEST_NAME> [OPTIONS + REQUEST_ITENS... ]"
                    ))
                    .about("Submit saved request")
                    .arg(Arg::new("inputs").value_name("REQUEST_NAME").required(true)),
            )
            .map(add_request_items_args)
            .map(add_raw_flag)
            .map(add_save_as_flag)
            .map(add_save_changes_to_current_request_flag)
            .map(add_manual_url_flag)
            .map(add_manual_method_flag)
            .unwrap(),
        )
        .subcommand(
            Some(
                Command::new("edit")
                    .override_usage(format!(
                        "treq run <REQUEST_NAME> [OPTIONS + REQUEST_ITENS... ]"
                    ))
                    .about("Edit saved request data, it does not submit")
                    .arg(Arg::new("inputs").value_name("REQUEST_NAME").required(true)),
            )
            .map(add_request_items_args)
            .map(add_raw_flag)
            .map(add_save_as_flag)
            .map(add_manual_url_flag)
            .map(add_manual_method_flag)
            .unwrap(),
        )
        .subcommand(
            Command::new("remove")
                .override_usage(format!("treq run <REQUEST_NAME> [OPTIONS]"))
                .about("Remove request")
                .arg(
                    Arg::new("inputs")
                        .value_name("inputs")
                        .required(true)
                        .num_args(1)
                        .help("Saved request to remove"),
                ),
        )
        .subcommand(
            Command::new("rename")
                .override_usage(format!(
                    "treq run <OLD_REQUEST_NAME> <NEW_REQUEST_NAME> [OPTIONS]"
                ))
                .about("Rename request")
                .arg(
                    Arg::new("inputs")
                        .value_name("inputs")
                        .required(true)
                        .num_args(2)
                        .help("All entrys"),
                ),
        )
        .subcommand(Command::new("ls").about("List all saved requests"))
        .subcommand(
            Command::new("inspect")
                .about("Show request details and datas")
                .arg(
                    Arg::new("inputs")
                        .value_name("REQUEST_NAME")
                        .required(true)
                        .num_args(1)
                        .help("Request name to inspect"),
                ),
        )
    };

    // Running without a subcommand
    app = Some(app.arg(Arg::new("inputs").value_name("URL")))
        .map(add_request_items_args)
        .map(add_raw_flag)
        .map(add_save_as_flag)
        .unwrap();

    app = app.override_usage(
        r#"
- Basic GET/POST request
  $ treq [URL] [OPTIONS + REQUEST_ITENS]

- Request with specific method
  $ treq [HTTP_METHOD] [URL] [OPTIONS + REQUEST_ITENS]

- Another commands
  $ treq [SUBCOMMAND] [INPUT] [OPTIONS + REQUEST_ITENS]
"#,
    );

    app = app.after_help(
        r#"
----------------------------------------
For more information, see 'treq --help'
----------------------------------------
>>>> Feel free to submit any issue or pull requests on our GitHub repository.
>>>> https://github.com/talis-fb/TReq
"#,
    );

    app = app.after_long_help(
        r#"
Examples
  Basic GET request
    $ treq example.com
    $ treq https://google.com

    # With explicit method
    $ treq GET example.com

  Requests with additional data
    # POST request with custom Content-Type header
    $ treq POST example.com Content-Type:application/json

    # Same POST request passing a JSON object { "language": "Rust", "food": "pizza" }
    $ treq POST example.com language=Rust food=pizza

    # Using multiples datas together
    $ treq POST example.com Content-Type:application/json language=Rust
    $ treq POST example.com Authorization:None name="John Doe" language=Rust

  Saving requests
    # After requesting you can save it to do the same request later
    $ treq POST example.com name="John Doe" --save-as main-endpoint
    $ treq run main-endpoint

    # You can also edit the fields and insert new datas in each submit
    $ treq run main-endpoint name="Jane" another-field="value"
    $ treq run main-endpoint Authorization:None 

    # Or save it as a new request
    $ treq run main-endpoint job="dev" --save-as endpoint-with-job

    # You can also save the changes in same request after submit
    $ treq run main-endpoint name="Peter" --save

    # Or just edit request data without submit
    $ treq edit main-endpoint name="Michael" job="dev"

    # For more complex data, you can use JSON object directly with `--raw`
    $ treq run example.com --raw '{ names: ["John", "Doe"] }' 'Content-Type:application/json'

----------------------------------------
>>>> Feel free to submit any issue or pull requests on our GitHub repository.
>>>> https://github.com/talis-fb/TReq
"#
        .trim(),
    );

    app
}

fn add_request_items_args(command: Command) -> Command {
    command.arg(
        Arg::new("request-items")
            .value_name("REQUEST_ITENS")
            .num_args(1..)
            .help(
                r#"
Optional key-value pairs to be included in the request.
Like Header values, in data to be serialized in JSON at payload.
    Use `--help` for more details"#
                    .trim(),
            )
            .long_help(
                r#"
Optional key-value pairs to be included in the request. The separator used determines the type of:
    HTTP header: ':'
      Content-Type:application/json
      User-Agent:bacon/1.0
      Accept:application/json
      Accept-Language:en-US

    Body data fields be serialized into a JSON object: '='
      name=John 
      language=Rust
      country=Brazil
      description='The best CLI HTTP client'"#
                    .trim(),
            ),
    )
}

fn add_raw_flag(command: Command) -> Command {
    command.arg(
        Arg::new("raw")
            .long("raw")
            .value_name("RAW_PAYLOD")
            .help("Raw payload value to be used (This exclude REQUEST_ITENS)"),
    )
}

fn add_save_as_flag(command: Command) -> Command {
    command.arg(
        Arg::new("save-as")
            .long("save-as")
            .value_name("SAVE_NAME")
            .help("Save builded request as <SAVE_NAME>, a named request"),
    )
}

fn add_save_changes_to_current_request_flag(command: Command) -> Command {
    command.arg(
        Arg::new("save")
            .long("save")
            .action(ArgAction::SetTrue)
            .help("Before submit request, save changes"),
    )
}

fn add_global_utils_flag(command: Command) -> Command {
    command.arg(
        Arg::new("no-color")
            .long("no-color")
            .action(ArgAction::SetTrue)
            .help("Print output in terminal without colors (default behavior if piped)"),
    )
}

fn add_manual_method_flag(command: Command) -> Command {
    command.arg(
        Arg::new("method-manual")
            .long("method")
            .value_name("METHOD_MANUAL")
            .value_parser(["GET", "POST", "PUT", "DELETE", "HEAD", "PATCH"])
            .ignore_case(true)
            .help("Set the HTTP Method when is not possible by subcommand"),
    )
}
fn add_manual_url_flag(command: Command) -> Command {
    command.arg(
        Arg::new("url-manual")
            .short('u')
            .long("url")
            .value_name("URL")
            .help("Sets a url manual"),
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_basic_get_request() {
        let root_matches = root_command()
            .try_get_matches_from(vec!["treq", "GET", "https://httpbin.org/get"])
            .unwrap();

        assert_eq!(None, root_matches.get_one::<String>("inputs"));
        assert_eq!(None, root_matches.get_one::<String>("request-items"));

        let (name_subcommand, matches_subcommand) = root_matches.subcommand().unwrap();
        assert_eq!("GET", name_subcommand);
        assert!(matches_subcommand.args_present());

        assert_eq!(
            vec!["https://httpbin.org/get"],
            matches_subcommand
                .get_many::<String>("inputs")
                .unwrap()
                .collect::<Vec<_>>(),
        );
    }

    #[test]
    fn test_basic_get_request_with_request_items() {
        let root_matches = root_command()
            .try_get_matches_from(vec![
                "treq",
                "GET",
                "https://httpbin.org/get",
                "Hello=World",
                "Content:json",
                "job=dev",
            ])
            .unwrap();

        let (_, subcommand) = root_matches.subcommand().unwrap();

        assert_eq!(
            vec!["Hello=World", "Content:json", "job=dev"],
            subcommand
                .get_many::<String>("request-items")
                .unwrap()
                .collect::<Vec<_>>(),
        );
    }

    #[test]
    fn test_basic_post_request() {
        let root_matches = root_command()
            .try_get_matches_from(vec![
                "treq",
                "POST",
                "https://httpbin.org/post",
                "--save-as",
                "test.json",
            ])
            .unwrap();

        assert_eq!(None, root_matches.get_one::<String>("inputs"));

        let (name_subcommand, matches_subcommand) = root_matches.subcommand().unwrap();
        assert_eq!("POST", name_subcommand);
        assert!(matches_subcommand.args_present());
    }

    #[test]
    fn test_basic_defatult_request() {
        let root_matches = root_command()
            .try_get_matches_from(vec![
                "treq",
                "https://httpbin.org/get",
                "--save-as",
                "test.json",
            ])
            .unwrap();

        assert_eq!(
            Some(&"https://httpbin.org/get".to_string()),
            root_matches.get_one::<String>("inputs"),
        );
        assert_eq!(None, root_matches.get_one::<String>("request-items"));
        assert_eq!(None, root_matches.subcommand());
    }

    #[test]
    fn test_basic_default_request_with_request_items() {
        let root_matches = root_command()
            .try_get_matches_from(vec![
                "treq",
                "https://httpbin.org/post",
                "Hello=World",
                "Content:json",
                "job=dev",
            ])
            .unwrap();

        assert_eq!(
            Some(&"https://httpbin.org/post".to_string()),
            root_matches.get_one::<String>("inputs"),
        );
        assert_eq!(
            vec!["Hello=World", "Content:json", "job=dev"],
            root_matches
                .get_many::<String>("request-items")
                .unwrap()
                .collect::<Vec<_>>(),
        );
        assert_eq!(root_matches.subcommand(), None);
    }
}
