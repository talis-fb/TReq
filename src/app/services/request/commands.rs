use super::commands::CommandRequestService as Command;
use super::service::RequestServiceInstance;
use crate::utils::commands::CommandClosureType;

pub type CommandRequestService = CommandClosureType<RequestServiceInstance>;

pub struct CommandsFactory;

impl CommandsFactory {
    pub fn do_nothing() -> Command {
        Box::new(|service| Ok(service))
    }
}

pub mod crud;
pub mod history;
