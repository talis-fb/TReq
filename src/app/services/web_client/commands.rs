use tokio::sync::oneshot::{self, Receiver};

// use super::commands::CommandRequestService as Command;
use super::{entity::Response, service::WebClientInstance};
use crate::app::services::request::entity::RequestData;
use crate::utils::commands::CommandClosureType;

pub type CommandWebClient = CommandClosureType<WebClientInstance>;

pub struct CommandsFactory;

impl CommandsFactory {
    pub fn do_nothing() -> CommandWebClient {
        Box::new(|service| Ok(service))
    }

    pub fn submit(request: RequestData) -> (CommandWebClient, Receiver<Result<Response, String>>) {
        let (tx, rx) = oneshot::channel();
        (
            Box::new(move |mut service| {
                let id_task = service.submit(request);
                let gg = service.get_task_request(id_task);

                if let Some(task) = gg {
                    tokio::task::spawn(async move {
                        let response = task.await.map_err(|e| e.to_string());

                        match response {
                            Ok(response) => tx.send(response),
                            Err(err) => tx.send(Err(err)),
                        }
                    });
                }
                Ok(service)
            }),
            rx,
        )
    }
}
