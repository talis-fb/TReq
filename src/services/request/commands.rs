use crate::services::request::facade::RequestServiceFacade;
use crate::utils::commands::Command as CommandAlias;

pub type CommandRequestService = CommandAlias<Box<dyn RequestServiceFacade>>;

pub struct Commands;

pub mod crud_request;
