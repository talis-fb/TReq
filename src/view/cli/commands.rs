#![allow(unused_variables)]
use std::io::{stderr, stdout};

use async_trait::async_trait;
use serde::Serialize;

use self::inspect_request::InspectRequestExecutor;
use self::save_request::SaveRequestExecutor;
use self::save_request_with_base_request::SaveRequestWithBaseRequestExecutor;
use self::show_list_all_request::ShowListAllRequestExecutor;
use self::submit_request::BasicRequestExecutor;
use self::submit_saved_request::SubmitSavedRequestExecutor;
use crate::app::backend::Backend;
use crate::app::services::request::entities::{OptionalRequestData, RequestData};
use crate::view::cli::output::writer::CrosstermCliWriter;

pub mod inspect_request;
pub mod save_request;
pub mod save_request_with_base_request;
pub mod show_list_all_request;
pub mod submit_request;
pub mod submit_saved_request;

#[async_trait]
pub trait CliCommand {
    async fn execute(self: Box<Self>, provider: &mut dyn Backend) -> anyhow::Result<()>;
}

impl<T: CliCommand + 'static> From<T> for Box<dyn CliCommand> {
    fn from(code: T) -> Self {
        Box::new(code)
    }
}

#[derive(Debug, PartialEq, Eq, Serialize)]
pub enum CliCommandChoice {
    SubmitRequest {
        request: RequestData,
    },

    SubmitSavedRequest {
        request_name: String,
        request_data: OptionalRequestData,
    },

    SaveRequest {
        request_name: String,
        request_data: OptionalRequestData,
        check_exists_before: bool,
    },
    SaveRequestWithBaseRequest {
        request_name: String,
        base_request_name: String,
        request_data: OptionalRequestData,
        check_exists_before: bool,
    },

    RemoveSavedRequest {
        request_name: String,
    },

    RenameSavedRequest {
        request_name: String,
        new_name: String,
    },

    ShowRequests,
    InspectRequest {
        request_name: String,
    },
}

pub fn get_executor_of_cli_command(command: CliCommandChoice) -> Box<dyn CliCommand> {
    let writer_stdout = CrosstermCliWriter::from(stdout());
    let writer_stderr = CrosstermCliWriter::from(stderr());

    match command {
        CliCommandChoice::SubmitRequest { request } => BasicRequestExecutor {
            request,
            writer_stdout,
            writer_stderr,
        }
        .into(),
        CliCommandChoice::SubmitSavedRequest {
            request_name,
            request_data,
        } => SubmitSavedRequestExecutor {
            request_name,
            request_data,
            writer_stdout,
            writer_stderr,
        }
        .into(),

        CliCommandChoice::SaveRequest {
            request_name,
            request_data,
            check_exists_before,
        } => SaveRequestExecutor {
            request_name,
            request_data,
            check_exists_before,
            writer_stdout,
            writer_stderr,
        }
        .into(),
        CliCommandChoice::SaveRequestWithBaseRequest {
            request_name,
            base_request_name,
            request_data,
            check_exists_before,
        } => SaveRequestWithBaseRequestExecutor {
            request_name,
            base_request_name,
            request_data,
            check_exists_before,
            writer_stdout,
            writer_stderr,
        }
        .into(),

        CliCommandChoice::ShowRequests => ShowListAllRequestExecutor {
            writer: writer_stdout,
        }
        .into(),
        CliCommandChoice::InspectRequest { request_name } => InspectRequestExecutor {
            request_name,
            writer: writer_stdout,
        }
        .into(),

        CliCommandChoice::RenameSavedRequest {
            request_name,
            new_name,
        } => todo!(),
        CliCommandChoice::RemoveSavedRequest { request_name } => todo!(),
    }
}
