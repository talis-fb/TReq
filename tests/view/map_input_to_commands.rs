#![allow(non_snake_case)]

use insta::assert_yaml_snapshot as assert_snapshot;
use treq::view::commands::ViewCommandChoice;
use treq::view::input::cli_definition::root_command;
use treq::view::input::cli_input::CliInput;
use treq::view::input_to_commands::map_input_to_commands;

fn process<'a>(input: impl IntoIterator<Item = &'a str>) -> anyhow::Result<Vec<ViewCommandChoice>> {
    let matches = root_command().get_matches_from(input);
    let inputs = CliInput::from_clap_matches(&matches)?;
    let commands_choices = map_input_to_commands(inputs)?;
    Ok(commands_choices)
}

#[test]
fn should_parse_to_normal_GET_submit_without_passing_method_as_subcommand_and_no_body() {
    let input = ["treq", "url.com"];
    let output = process(input).unwrap();
    assert!(output.len() == 1);
    assert_snapshot!(output);
}

#[test]
fn should_parse_to_normal_POST_submit_without_passing_method_as_subcommand_but_passing_some_body_data(
) {
    let input = ["treq", "url.com", "Hello=World"];
    let output = process(input).unwrap();
    assert!(output.len() == 1);
    assert_snapshot!(output);
}

#[test]
fn should_ignore_body_inputs_in_GET_request() {
    let input = ["treq", "GET", "url.com", "Hello=World"];
    let output = process(input).unwrap();
    assert!(output.len() == 1);
    assert_snapshot!(output);
}

#[test]
fn should_parse_all_methods_subcommands_to_normal_submits() {
    let all_methods = ["GET", "POST", "PUT", "DELETE", "HEAD", "PATCH"];

    let inputs = all_methods
        .into_iter()
        .map(|method| ["treq", method, "url.com"])
        .collect::<Vec<_>>();

    inputs.into_iter().for_each(|input| {
        let output = process(input).unwrap();
        debug_assert!(output.len() == 1, "{:?}", output);
        assert_snapshot!(output);
    });
}

#[test]
fn should_parse_same_way_with_or_without_protocol_in_url() {
    let input1 = ["treq", "url.com"];
    let input2 = ["treq", "http://url.com"];
    process(input1).unwrap();
    process(input2).unwrap();
}

#[test]
fn should_error_if_no_input() {
    let input = ["treq"];
    let output = process(input);
    assert!(output.is_err());
}

#[test]
fn should_raw_flag_work_equal_param_body_definition() {
    let input1 = ["treq", "POST", "url.com", "Hello=World"];
    let input2 = ["treq", "POST", "url.com", "--raw", r#"{"Hello":"World"}"#];
    let output1 = process(input1);
    let output2 = process(input2);

    debug_assert!(output1.is_ok(), "{:?}", output1);
    debug_assert!(output2.is_ok(), "{:?}", output2);
    assert_eq!(output1.unwrap(), output2.unwrap());
}

#[test]
fn should_merge_inputs_of_raw_flag_and_param_body() {
    let input = [
        "treq",
        "POST",
        "url.com",
        "--raw",
        r#"{"name":"Thales"}"#,
        "age=40",
        "job=Dev",
    ];
    let output = process(input);
    debug_assert!(output.is_ok(), "{:?}", output);
    assert_snapshot!(output.unwrap());
}

#[test]
fn should_set_query_params_by_request_items() {
    // Url final => url.com/?search=Rust&country=br
    let input = ["treq", "GET", "url.com", "search==Rust", "country==br"];
    let output = process(input);
    debug_assert!(output.is_ok(), "{:?}", output);
    assert_snapshot!(output.unwrap());
}

#[test]
fn should_execute_with_valid_urls() {
    let VALID_URLS = [
        "google.com",
        "google.com/",
        "google.com?",
        "google.com:81",
        "google.com:81/",
        "google.com/search/advanced",
        "google.com/search/advanced/",
        "google.com?search=Rust",
        "google.com?search=Rust&country=br",
        "google.com/search/advanced?name=john",
        "google.com/search/advanced/?name=john",
        "google.com/search/advanced?name=john&sort=true",
        "google.com/search/advanced?name=john&sort=true#landing-page",
        "google.com/search/advanced#landing-page",
        "google.com/search/advanced/#landing-page",
        "google.com#landing-page",
        "google.com/#landing-page",
        "maps.google.com/#landing-page",
        "subdomain.from-subdomain.maps.google.com/#",
        "subdomain.from-subdomain.maps.google.com:9999/#",
        "localhost",
        "localhost/",
        "localhost:8081/",
        "localhost:8081/api/v1/local",
    ];

    for url in VALID_URLS {
        let input = ["treq", "GET", url];
        let output = process(input);
        debug_assert!(output.is_ok(), "{:?}", output);
    }
}

// ---------------------------------
// Validators    -------------------
// ---------------------------------
#[test]
fn should_validate_alias_for_localhost_requests() {
    let input = ["treq", "GET", ":8080", "search==Rust", "country==br"];
    let output = process(input);
    debug_assert!(output.is_ok(), "{:?}", output);
    assert_snapshot!(output.unwrap());
}

#[test]
fn should_validate_alias_for_localhost_requests_with_manual_url() {
    let input = [
        "treq",
        "run",
        "some-request",
        "search==Rust",
        "--url",
        ":8080",
    ];
    let output = process(input);
    debug_assert!(output.is_ok(), "{:?}", output);
    assert_snapshot!(output.unwrap());
}

#[test]
fn should_validate_if_error_on_raw_value_as_not_map_and_body_insert() {
    // Ok
    let input = [
        "treq",
        "GET",
        "url.com",
        "--raw",
        "{\"Hello\":\"World\"}",
        "name=John",
    ];
    let output = process(input);
    debug_assert!(output.is_ok(), "{:?}", output);

    // Ok
    let input = ["treq", "GET", "url.com", "--raw", "{}", "name=John"];
    let output = process(input);
    debug_assert!(output.is_ok(), "{:?}", output);

    // Err
    let input = ["treq", "GET", "url.com", "--raw", "some-value", "name=John"];
    let output = process(input);
    debug_assert!(output.is_err(), "{:?}", output);
    debug_assert!(
        output
            .as_ref()
            .unwrap_err()
            .to_string()
            .contains("raw body"),
        "{:?}",
        output
    );

    // Err
    let input = [
        "treq",
        "GET",
        "url.com",
        "--raw",
        "[ \"Hello\",\"World\" ]",
        "name=John",
    ];
    let output = process(input);
    debug_assert!(output.is_err(), "{:?}", output);
    debug_assert!(
        output
            .as_ref()
            .unwrap_err()
            .to_string()
            .contains("raw body"),
        "{:?}",
        output
    );
}
