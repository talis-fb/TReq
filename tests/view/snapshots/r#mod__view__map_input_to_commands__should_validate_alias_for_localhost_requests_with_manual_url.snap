---
source: tests/view/map_input_to_commands.rs
expression: output.unwrap()
---
- SubmitSavedRequest:
    request_name: some-request
    request_data:
      url:
        ValidatedUrl:
          protocol: ~
          host: localhost
          port: 8080
          paths: []
          query_params:
            - - search
              - Rust
          anchor: ~
      method: ~
      headers: ~
      body: ~
    view_options:
      print_body_only: false
      suppress_output: false
